### PR TITLE
workflows: Implement ExecuteWorkflow

### PR DESCRIPTION
**Pull request by bduffany**
The UI can use this RPC to allow manually re-running a workflow. (The UI has all the information that is needed already.)

Because we don't implement `IsPrivate` for non-GitHub repos (for now), the UI will only show this button for GitHub repos.

---

**Version bump**: Patch

### CODE REVIEW

**Comment by siggisim** [comment]
> @siggisim mostly looking for your input on the planned UI impl in the PR description

Can we insert a bare-bones invocation into the DB as part of the handler (similar to when a bazel build is in the "analysis phase")?

If for some reason the invocation fails to be created - we'll just refresh the UI forever and never give any feedback to the user.

**Comment by siggisim** [comment]
Also not sure what the repo being private has to do with any of this?

**Comment by bduffany** [comment]
@siggisim I like the idea of inserting the bare-bones invocation -- @tylerwilliams wdyt?

Also, whether the invocation is private determines whether we pass secrets to the CI runner (in particular the GitHub access token).

File: enterprise/server/cmd/ci_runner/main.go [hunk 1]
```diff
@@ -141,13 +143,26 @@ func RunAllActions(ctx context.Context, cfg *config.BuildBuddyConfig, im *initMe
...
 
+		iid := ""
+		// Respect the invocation ID flag only when running a single action
+		// (via ExecuteWorkflow).
+		if *actionName != "" {
+			iid = *invocationID
+		}
+		if iid == "" {
+			iid = newUUID()
+		}
```
Comment (tylerwilliams) [inline enterprise/server/cmd/ci_runner/main.go:163]: Why is this necessary?

EDIT: OK, I guess it's because you want to ensure 1-1 mapping between actions and invocation IDs. Then why not return an error or log.Fatalf in this case since it indicates bad input?

File: enterprise/server/cmd/server/main.go [hunk 2]
```diff
@@ -237,6 +239,13 @@ func main() {
...
 		}
 		realEnv.SetSchedulerService(schedulerServer)
+
+		// Fulfill internal remote execution requests locally.
+		conn, err := grpc_client.DialTarget(fmt.Sprintf("grpc://localhost:%d", *libmain.GRPCPort))
+		if err != nil {
+			log.Fatalf("Error initializing remote execution client: %s", err)
+		}
+		realEnv.SetRemoteExecutionClient(repb.NewExecutionClient(conn))
 	}
```
Comment (tylerwilliams) [inline enterprise/server/cmd/server/main.go:249]: does this work correctly if remote execution is not configured?

File: enterprise/server/workflow/service/service.go [hunk 3]
```diff
@@ -268,6 +274,142 @@ func (ws *workflowService) GetWorkflows(ctx context.Context, req *wfpb.GetWorkfl
...
+		return nil, err
+	}
+	invocationID := invocationUUID.String()
+	extraCIRunnerArgs := []string{
+		fmt.Sprintf("--action_name=%s", req.GetActionName()),
+		fmt.Sprintf("--invocation_id=%s", invocationID),
+	}
+
+	executionID, err := ws.executeWorkflow(ctx, wf, wd, extraCIRunnerArgs)
+
```
Comment (tylerwilliams) [inline enterprise/server/workflow/service/service.go:349]: rm newline

File: enterprise/server/workflow/service/service.go [hunk 4]
```diff
@@ -268,6 +274,142 @@ func (ws *workflowService) GetWorkflows(ctx context.Context, req *wfpb.GetWorkfl
...
+	// Authenticate
+	user, err := perms.AuthenticatedUser(ctx, ws.env)
+	if err != nil {
+		return nil, err
+	}
+
+	// Lookup workflow
+	wf := &tables.Workflow{}
+	db := ws.env.GetDBHandle()
+	err = db.Raw(
```
Comment (tylerwilliams) [inline enterprise/server/workflow/service/service.go:301]: can you use:

 err := env.GetDBHandle().Transaction(ctx, func(tx *db.DB) error {

instead please?

File: enterprise/server/workflow/service/service.go [hunk 5]
```diff
@@ -268,6 +274,142 @@ func (ws *workflowService) GetWorkflows(ctx context.Context, req *wfpb.GetWorkfl
...
+		`SELECT group_id, repo_url, access_token, perms FROM Workflows WHERE workflow_id = ?`,
+		req.GetWorkflowId(),
+	).Take(wf).Error
+	if err != nil {
+		if errors.Is(err, gorm.ErrRecordNotFound) {
```
Comment (tylerwilliams) [inline enterprise/server/workflow/service/service.go:306]: nit: please use IsRecordNotFound

(and remove any gorm imports)

File: enterprise/server/workflow/service/service.go [hunk 6]
```diff
@@ -268,6 +274,142 @@ func (ws *workflowService) GetWorkflows(ctx context.Context, req *wfpb.GetWorkfl
...
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case err := <-errCh:
+			return err
+		case op := <-opCh:
+			stage = operation.ExtractStage(op)
+		case <-time.After(1 * time.Second):
+			// Fall through
```
Comment (tylerwilliams) [inline enterprise/server/workflow/service/service.go:396]: is this the same as having a break here? I think that would be more explicit

File: enterprise/server/workflow/service/service.go [hunk 7]
```diff
@@ -268,6 +274,142 @@ func (ws *workflowService) GetWorkflows(ctx context.Context, req *wfpb.GetWorkfl
...
+		}
+		if stage == repb.ExecutionStage_EXECUTING || stage == repb.ExecutionStage_COMPLETED {
+			_, err := indb.LookupInvocation(ctx, invocationID)
+			if err == nil {
+				return nil
+			}
+			if !db.IsRecordNotFound(err) {
+				return err
+			}
```
Comment (tylerwilliams) [inline enterprise/server/workflow/service/service.go:405]: kinda cheeky, but could to shorten this to:

if _, err := indb.LookupInvocation(ctx, invocationID); !db.IsRecordNotFound(err) {
  return err
}

File: enterprise/server/workflow/service/service.go [hunk 8]
```diff
@@ -416,6 +561,9 @@ func (ws *workflowService) createActionForWorkflow(ctx context.Context, wf *tabl
 		DoNotCache:      true,
 	}
 	actionDigest, err := cachetools.UploadProtoToCAS(ctx, cache, instanceName, action)
+	if err != nil {
+		return nil, err
+	}
 	return actionDigest, err
```
Comment (tylerwilliams) [inline enterprise/server/workflow/service/service.go:567]: I don't fully understand this change ^. Why not:

return cachetools.UploadProtoToCAS(ctx, cache, instanceName, action)

**Review by tylerwilliams — CHANGES_REQUESTED** [review]

File: enterprise/server/cmd/ci_runner/main.go (hunk 1 above)
Comment (bduffany) [reply enterprise/server/cmd/ci_runner/main.go:163]: Done (checked the flags at the start of `main`)

File: enterprise/server/cmd/server/main.go (hunk 2 above)
Comment (bduffany) [reply enterprise/server/cmd/server/main.go:249]: It's hard to tell on GitHub but this code is within a conditional that checks whether remote execution is configured. For good measure I also added a nil check on executionClient within `waitForWorkflowInvocationCreated` and return UnimplementedError if so.

File: enterprise/server/workflow/service/service.go (hunk 4 above)
Comment (bduffany) [reply enterprise/server/workflow/service/service.go:301]: Hm, why is a transaction needed if there's only a single query executed?

File: enterprise/server/workflow/service/service.go (hunk 5 above)
Comment (bduffany) [reply enterprise/server/workflow/service/service.go:306]: Done

File: enterprise/server/workflow/service/service.go (hunk 3 above)
Comment (bduffany) [reply enterprise/server/workflow/service/service.go:349]: Done

File: enterprise/server/workflow/service/service.go (hunk 6 above)
Comment (bduffany) [reply enterprise/server/workflow/service/service.go:396]: Yep, done

File: enterprise/server/workflow/service/service.go (hunk 7 above)
Comment (bduffany) [reply enterprise/server/workflow/service/service.go:405]: I prefer the explicit nil check + return nil because otherwise the reader has to do a bit of extra thinking to convince themselves that this function can ever return `nil`.

File: enterprise/server/workflow/service/service.go (hunk 8 above)
Comment (bduffany) [reply enterprise/server/workflow/service/service.go:567]: Removed the extraneous `err != nil` check. But I like the existing `actionDigest, err := ...` followed by `return actionDigest, err` because it makes the code more self-documenting.

**Review by tylerwilliams — APPROVED** [review]
LGTM
